services/resource-usage-tracker/tests/unit/with_dbs/test_process_rabbitmq_message_with_billing.py

```diff
@@ -0,0 +1,149 @@
+    "adminer",
+]
+
+
+@pytest.fixture()
+def resource_tracker_pricing_tables_db(postgres_db: sa.engine.Engine) -> Iterator[None]:
+    with postgres_db.connect() as con:
+        con.execute(
+            resource_tracker_pricing_plans.insert().values(
+                product_name="osparc",
+                name="test_name",
+                description="",
+                classification="TIER",
+                is_active=True,
+            )
+        )
+        con.execute(
+            resource_tracker_pricing_details.insert().values(
+                pricing_plan_id=1,
+                unit_name="S",
+                cost_per_unit=500,
+                valid_from=datetime.now(tz=timezone.utc),
+            ),
+            simcore_default=True,
+            specific_info={},
+        )
+        con.execute(
+            resource_tracker_pricing_details.insert().values(
+                pricing_plan_id=1,
+                unit_name="M",
+                cost_per_unit=1000,
+                valid_from=datetime.now(tz=timezone.utc),
+            ),
+            simcore_default=False,
+            specific_info={},
+        )
+        con.execute(
+            resource_tracker_pricing_details.insert().values(
+                pricing_plan_id=1,
+                unit_name="L",
+                cost_per_unit=1500,
+                valid_from=datetime.now(tz=timezone.utc),
+            ),
+            simcore_default=False,
+            specific_info={},
+        )
+        con.execute(
+            resource_tracker_pricing_plan_to_service.insert().values(
+                pricing_plan_id=1,
+                product="osparc",
+                service_key="simcore/services/comp/itis/sleeper",
+                service_version="1.0.16",
+            )
+        )
+
+        yield
+
+        con.execute(resource_tracker_pricing_plan_to_service.delete())
+        con.execute(resource_tracker_pricing_details.delete())
+        con.execute(resource_tracker_pricing_plans.delete())
+
+
+@pytest.mark.testit
+async def test_process_event_functions(
+    mocked_setup_rabbitmq,
+    random_rabbit_message_start,
+    mocked_redis_server: None,
+    postgres_db: sa.engine.Engine,
+    resource_tracker_service_run_db,
+    resource_tracker_pricing_tables_db,
+    initialized_app,
+):
+    engine = initialized_app.state.engine
+
+    msg = random_rabbit_message_start(
+        wallet_id=1, wallet_name="test", pricing_plan_id=1, pricing_detail_id=1
+    )
+    resource_tacker_repo: ResourceTrackerRepository = ResourceTrackerRepository(
+        db_engine=engine
+    )
+    await _process_start_event(resource_tacker_repo, msg)
+    output = await assert_credit_transactions_db_row(postgres_db, msg.service_run_id)
+    assert float(output[8]) == 0.0
+    assert output[9] == "PENDING"
+    assert output[10] == "DEDUCT_SERVICE_RUN"
+    first_occurence_of_last_heartbeat_at = output[14]
+    modified_at = output[15]
+
+    await asyncio.sleep(2)
```

Comment:
This one is not even needed there (and tenacity is not needed at all I am not waiting for anything) it is just so there is some time between these events so there is some visual change in the created_at message timestamp which is nice during testing/debugging.

Reply:
But generally I will improve the tests in next iterations

Reply:
if this is just about ensuring the asyncio event loop refreshes then just do ```await asyncio.sleep(0)```. Otherwise, please name the 2 into a variable that explains the why it is in there. After all the flaky shit we went through in the tests, I think sleeps are very suspicious to say the least ;)

Reply:
other wise, create an _assert fct that uses tenacity and waits for whatever it is you want to see change to actually change.